packages/backend/src/api/controllers/events/EventsController.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  ApiEvents,
+  EventChart,
+  EventChartPoint,
+  ProjectId,
+  UnixTime,
+} from '@l2beat/types'
+
+import { EventUpdater } from '../../../core/events/EventUpdater'
+import { ProjectInfo } from '../../../model'
+import { EventRepository } from '../../../peripherals/database/EventRepository'
+import {
+  getHourlyMinTimestamp,
+  getSixHourlyMinTimestamp,
+} from '../report/charts'
+import { renderShowcasePage } from './ShowcasePage'
+
+export class EventsController {
+  constructor(
+    private eventsRepository: EventRepository,
+    private eventUpdater: EventUpdater,
+    private projects: ProjectInfo[],
+  ) {}
+
+  async getEvents(): Promise<ApiEvents> {
+    const main: ApiEvents = {
+      projects: {},
+    }
+
+    const timestamp = this.eventUpdater.getLastProcessed()
+
+    const config: {
+      granularity: 'hourly' | 'sixHourly' | 'daily'
+      timestamp: UnixTime | undefined
+    }[] = [
+      { granularity: 'hourly', timestamp: getHourlyMinTimestamp(timestamp) },
+      {
+        granularity: 'sixHourly',
+        timestamp: getSixHourlyMinTimestamp(timestamp),
+      },
+      { granularity: 'daily', timestamp: undefined },
+    ]
+
+    await Promise.all(
+      this.projects.map(async ({ projectId, events }) => {
+        const eventNames = events.map((e) => e.name)
+
+        const eventCharts = await Promise.all(
+          config.map((c) =>
+            this.getEventChart(
```

Comment:
I would probably call this.eventsRepository.getByProject directly and transform the result in a pure function that's easy to test.